File: MiddleClick/Controller.m
#import "Controller.h"
#include "TrayMenu.h"
#import <Cocoa/Cocoa.h>
#include <CoreFoundation/CoreFoundation.h>
#import <Foundation/Foundation.h>
#include <math.h>
#include <unistd.h>

#pragma mark Multitouch API

typedef struct {
  float x, y;
} mtPoint;
typedef struct {
  mtPoint pos, vel;
} mtReadout;

typedef struct {
  int frame;
  double timestamp;
  int identifier, state, foo3, foo4;
  mtReadout normalized;
  float size;
  int zero1;
  float angle, majorAxis, minorAxis; // ellipsoid
  mtReadout mm;
  int zero2[2];
  float unk2;
} Finger;

typedef void* MTDeviceRef;
typedef int (*MTContactCallbackFunction)(int, Finger*, int, double, int);
MTDeviceRef MTDeviceCreateDefault(void);
CFMutableArrayRef MTDeviceCreateList(void);
void MTDeviceRelease(MTDeviceRef);
void MTRegisterContactFrameCallback(MTDeviceRef, MTContactCallbackFunction);
void MTUnregisterContactFrameCallback(MTDeviceRef, MTContactCallbackFunction);
void MTDeviceStart(MTDeviceRef, int); // thanks comex
void MTDeviceStop(MTDeviceRef);

#pragma mark Globals

NSDate* touchStartTime;
float middleclickX, middleclickY;
float middleclickX2, middleclickY2;

BOOL needToClick;
long fingersQua;
BOOL threeDown;
BOOL maybeMiddleClick;
BOOL wasThreeDown;
NSMutableArray* currentDeviceList;
CFMachPortRef currentEventTap;
CFRunLoopSourceRef currentRunLoopSource;

#pragma mark Implementation

@implementation Controller {
  NSTimer* _restartTimer __weak; // Using `weak` so that the pointer is automatically set to `nil` when the referenced object is released ( https://en.wikipedia.org/wiki/Automatic_Reference_Counting#Zeroing_Weak_References ). This helps preventing fatal EXC_BAD_ACCESS.
}

- (void)start
{
  NSLog(@"Starting all listeners...");

  threeDown = NO;
  wasThreeDown = NO;
  
  fingersQua = [[NSUserDefaults standardUserDefaults] integerForKey:@"fingers"];
  
  needToClick =
  [[NSUserDefaults standardUserDefaults] boolForKey:@"needClick"];
  
  NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
  [NSApplication sharedApplication];
  
  registerTouchCallback();
  
  // register a callback to know when osx come back from sleep
  [[[NSWorkspace sharedWorkspace] notificationCenter]
   addObserver:self
   selector:@selector(receiveWakeNote:)
   name:NSWorkspaceDidWakeNotification
   object:NULL];
  
  // Register IOService notifications for added devices.
  IONotificationPortRef port = IONotificationPortCreate(kIOMasterPortDefault);
  CFRunLoopAddSource(CFRunLoopGetMain(),
                     IONotificationPortGetRunLoopSource(port),
                     kCFRunLoopDefaultMode);
  io_iterator_t handle;
  kern_return_t err = IOServiceAddMatchingNotification(
                                                       port, kIOFirstMatchNotification,
                                                       IOServiceMatching("AppleMultitouchDevice"), multitouchDeviceAddedCallback,
                                                       self, &handle);
  if (err) {
    NSLog(@"Failed to register notification for touchpad attach: %xd, will not "
          @"handle newly "
          @"attached devices",
          err);
    IONotificationPortDestroy(port);
  } else {
    io_object_t item;
    while ((item = IOIteratorNext(handle))) {
      IOObjectRelease(item);
    }
  }
  
  // when displays are reconfigured restart of the app is needed, so add a calback to the
  // reconifguration of Core Graphics
  CGDisplayRegisterReconfigurationCallback(displayReconfigurationCallBack, self);
  
  [self registerMouseCallback:pool];
}

static void stopUnstableListeners()
{
    NSLog(@"Stopping unstable listeners...");

    unregisterTouchCallback();
    unregisterMouseCallback();
}

- (void)startUnstableListeners
{
  NSLog(@"Starting unstable listeners...");
    
  NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];

  registerTouchCallback();
  [self registerMouseCallback:pool];
}

static void registerTouchCallback()
{
    // Get list of all multi touch devices
    NSMutableArray* deviceList = (NSMutableArray*)MTDeviceCreateList(); // grab our device list
    currentDeviceList = deviceList;

    // Iterate and register callbacks for multitouch devices.
    for (int i = 0; i < [deviceList count]; i++) // iterate available devices
    {
      registerMTDeviceCallback((MTDeviceRef)[deviceList objectAtIndex:i], touchCallback);
    }
}
static void unregisterTouchCallback()
{
    // Get list of all multi touch devices
    NSMutableArray* deviceList = currentDeviceList; // grab our device list

    // Iterate and unregister callbacks for multitouch devices.
    for (int i = 0; i < [deviceList count]; i++) // iterate available devices
    {
      unregisterMTDeviceCallback((MTDeviceRef)[deviceList objectAtIndex:i], touchCallback);
    }
}

- (void)registerMouseCallback:(NSAutoreleasePool*)pool
{
    // we only want to see left mouse down and left mouse up, because we only want
    // to change that one
    CGEventMask eventMask = (CGEventMaskBit(kCGEventLeftMouseDown) | CGEventMaskBit(kCGEventLeftMouseUp));

    // create eventTap which listens for core grpahic events with the filter
    // sepcified above (so left mouse down and up again)
    CFMachPortRef eventTap = CGEventTapCreate(
                                              kCGHIDEventTap, kCGHeadInsertEventTap, kCGEventTapOptionDefault,
                                              eventMask, mouseCallback, NULL);
    currentEventTap = eventTap;

    if (eventTap) {
        // Add to the current run loop.
        CFRunLoopSourceRef runLoopSource = CFMachPortCreateRunLoopSource(kCFAllocatorDefault, eventTap, 0);
        currentRunLoopSource = runLoopSource;
        CFRunLoopAddSource(CFRunLoopGetCurrent(), runLoopSource,
                           kCFRunLoopCommonModes);

        // Enable the event tap.
        CGEventTapEnable(eventTap, true);

        // release pool before exit
        [pool release];
    } else {
        NSLog(@"Couldn't create event tap! Check accessibility permissions.");
        [[NSUserDefaults standardUserDefaults] setBool:1 forKey:@"NSStatusItem Visible Item-0"];
        [self scheduleRestart:5];
    }
}
static void unregisterMouseCallback()
{
    // Remove from the current run loop.
    if (currentRunLoopSource) {
        CFRunLoopRemoveSource(CFRunLoopGetCurrent(), currentRunLoopSource, kCFRunLoopCommonModes);
    }
    // Disable the event tap.
    if (currentEventTap) {
        CGEventTapEnable(currentEventTap, false);
    }
}

/// Schedule listeners to be restarted, if a restart is pending, delay it.
- (void)scheduleRestart:(NSTimeInterval)delay
{
  if (_restartTimer != nil) { // Check whether the timer object was not released.
    [_restartTimer invalidate]; // Invalidate any existing timer.
  }
  
  _restartTimer = [NSTimer scheduledTimerWithTimeInterval:delay
                                                  repeats:NO
                                                    block:^(NSTimer* timer) {
                                                      [self restartListeners];
                                                    }];
}

// Callback for system wake up. This restarts the app to initialize callbacks.
// Can be tested by entering `pmset sleepnow` in the Terminal
- (void)receiveWakeNote:(NSNotification*)note
{
  NSLog(@"System woke up, restarting...");
  [self scheduleRestart:10];
}

- (BOOL)getClickMode
{
  return needToClick;
}

- (void)setMode:(BOOL)click
{
  [[NSUserDefaults standardUserDefaults] setBool:click forKey:@"needClick"];
  needToClick = click;
}

// listening to mouse clicks to replace them with middle clicks if there are 3
// fingers down at the time of clicking this is done by replacing the left click
// down with a other click down and setting the button number to middle click
// when 3 fingers are down when clicking, and by replacing left click up with
// other click up and setting three button number to middle click when 3 fingers
// were down when the last click went down.
CGEventRef mouseCallback(CGEventTapProxy proxy, CGEventType type,
                         CGEventRef event, void* refcon)
{
  if (needToClick) {
    if (threeDown && type == kCGEventLeftMouseDown) {
      wasThreeDown = YES;
      CGEventSetType(event, kCGEventOtherMouseDown);
      CGEventSetIntegerValueField(event, kCGMouseEventButtonNumber,
                                  kCGMouseButtonCenter);
      threeDown = NO;
    }
    
    if (wasThreeDown && type == kCGEventLeftMouseUp) {
      wasThreeDown = NO;
      CGEventSetType(event, kCGEventOtherMouseUp);
      CGEventSetIntegerValueField(event, kCGMouseEventButtonNumber,
                                  kCGMouseButtonCenter);
    }
  }
  return event;
}

// mulittouch callback, see what is touched. If 3 are on the mouse set
// threedowns, else unset threedowns.
int touchCallback(int device, Finger* data, int nFingers, double timestamp,
                  int frame)
{
  NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
  fingersQua = [[NSUserDefaults standardUserDefaults] integerForKey:@"fingers"];
  
  if (needToClick) {
    if (nFingers == fingersQua) {
      if (!threeDown) {
        threeDown = YES;
      }
    }
    
    if (nFingers != fingersQua) {
      if (threeDown) {
        threeDown = NO;
      }
    }
  } else {
    if (nFingers == 0) {
      touchStartTime = NULL;
      if (middleclickX + middleclickY) {
        float delta = ABS(middleclickX - middleclickX2) + ABS(middleclickY - middleclickY2);
        if (delta < 0.4f) {
          // Emulate a middle click
          
          // get the current pointer location
          CGEventRef ourEvent = CGEventCreate(NULL);
          CGPoint ourLoc = CGEventGetLocation(ourEvent);
          
          CGEventPost(kCGHIDEventTap,
                      CGEventCreateMouseEvent(NULL, kCGEventOtherMouseDown,
                                              ourLoc, kCGMouseButtonCenter));
          CGEventPost(kCGHIDEventTap,
                      CGEventCreateMouseEvent(NULL, kCGEventOtherMouseUp,
                                              ourLoc, kCGMouseButtonCenter));
        }
      }
    } else if (nFingers > 0 && touchStartTime == NULL) {
      NSDate* now = [[NSDate alloc] init];
      touchStartTime = [now retain];
      [now release];
      
      maybeMiddleClick = YES;
      middleclickX = 0.0f;
      middleclickY = 0.0f;
    } else {
      if (maybeMiddleClick == YES) {
        NSTimeInterval elapsedTime = -[touchStartTime timeIntervalSinceNow];
        if (elapsedTime > 0.5f)
          maybeMiddleClick = NO;
      }
    }
    
    if (nFingers > fingersQua) {
      maybeMiddleClick = NO;
      middleclickX = 0.0f;
      middleclickY = 0.0f;
    }
    
    if (nFingers == fingersQua) {
      if (maybeMiddleClick == YES) {
        for (int i = 0; i < fingersQua; i++)
        {
          mtPoint pos = ((Finger *)&data[i])->normalized.pos;
          middleclickX += pos.x;
          middleclickY += pos.y;
        }
        middleclickX2 = middleclickX;
        middleclickY2 = middleclickY;
        maybeMiddleClick = NO;
      } else {
        middleclickX2 = 0.0f;
        middleclickY2 = 0.0f;
        for (int i = 0; i < fingersQua; i++)
        {
          mtPoint pos = ((Finger *)&data[i])->normalized.pos;
          middleclickX2 += pos.x;
          middleclickY2 += pos.y;
        }
      }
    }
  }
  
  [pool release];
  return 0;
}

/// Restart the listeners when devices are connected/invalidated.
- (void)restartListeners
{
  NSLog(@"Restarting app functionality...");
  stopUnstableListeners();
  [self startUnstableListeners];
}

/// Callback when a multitouch device is added.
void multitouchDeviceAddedCallback(void* _controller,
                                   io_iterator_t iterator)
{
  io_object_t item;
  while ((item = IOIteratorNext(iterator))) {
    IOObjectRelease(item);
  }
  
  NSLog(@"Multitouch device added, restarting...");
  Controller* controller = (Controller*)_controller;
  [controller scheduleRestart:2];
}

void displayReconfigurationCallBack(CGDirectDisplayID display, CGDisplayChangeSummaryFlags flags, void* _controller)
{
  if(flags & kCGDisplaySetModeFlag || flags & kCGDisplayAddFlag || flags & kCGDisplayRemoveFlag || flags & kCGDisplayDisabledFlag)
  {
    NSLog(@"Display reconfigured, restarting...");
    Controller* controller = (Controller*)_controller;
    [controller scheduleRestart:2];
  }
}

static void registerMTDeviceCallback(MTDeviceRef device, MTContactCallbackFunction callback) {
    MTRegisterContactFrameCallback(device, callback); // assign callback for device
    MTDeviceStart(device, 0); // start sending events
}
static void unregisterMTDeviceCallback(MTDeviceRef device, MTContactCallbackFunction callback) {
    MTUnregisterContactFrameCallback(device, callback); // unassign callback for device
    MTDeviceStop(device); // stop sending events
    MTDeviceRelease(device);
}

@end

File: MiddleClick/Controller.swift
import Cocoa
import CoreGraphics
import Foundation
import IOKit
import IOKit.hid

// Globals
@MainActor public var needToClick = MiddleClickConfig.needClickDefault
@MainActor public var threeDown = false
@MainActor public var wasThreeDown = false
@MainActor public var maybeMiddleClick = false
@MainActor public var touchStartTime: Date?
@MainActor public var middleClickX: Float = 0.0
@MainActor public var middleClickY: Float = 0.0
@MainActor public var middleClickX2: Float = 0.0
@MainActor public var middleClickY2: Float = 0.0
@MainActor public var fingersQua = UserDefaults.standard.integer(
  forKey: MiddleClickConfig.fingersNumKey)
@MainActor public let maxDistanceDelta = UserDefaults.standard.float(
  forKey: MiddleClickConfig.maxDistanceDeltaKey)
@MainActor public let maxTimeDelta =
UserDefaults.standard
  .double(forKey: MiddleClickConfig.maxTimeDeltaMsKey) / 1000.0
@MainActor public var allowMoreFingers = UserDefaults.standard.bool(
  forKey: MiddleClickConfig.allowMoreFingersKey)

@MainActor class Controller: NSObject {
  weak var restartTimer: Timer?
  var currentEventTap: CFMachPort?
  var currentRunLoopSource: CFRunLoopSource?

  static let fastRestart = false
  static let wakeRestartTimeout: TimeInterval = fastRestart ? 2 : 10

  static let immediateRestart = false

  var currentDeviceList: [MTDevice] = []

  func start() {
    NSLog("Starting all listeners...")

    needToClick =
      UserDefaults.standard.value(forKey: "needClick") as? Bool
      ?? getIsSystemTapToClickDisabled()

    registerTouchCallback()

    NSWorkspace.shared.notificationCenter.addObserver(
      self,
      selector: #selector(receiveWakeNote(_:)),
      name: NSWorkspace.didWakeNotification,
      object: nil
    )

    guard let port = IONotificationPortCreate(kIOMasterPortDefault) else {
      NSLog("Failed to create IONotificationPort.")
      return
    }

    if let runLoopSource = IONotificationPortGetRunLoopSource(port)?
      .takeUnretainedValue()
    {
      CFRunLoopAddSource(CFRunLoopGetMain(), runLoopSource, .defaultMode)
    } else {
      NSLog("Failed to get run loop source from IONotificationPort.")
      IONotificationPortDestroy(port)
      return
    }

    var handle: io_iterator_t = 0
    let err = IOServiceAddMatchingNotification(
      port,
      kIOFirstMatchNotification,
      IOServiceMatching("AppleMultitouchDevice"),
      multitouchDeviceAddedCallback,
      Unmanaged.passUnretained(self).toOpaque(),
      &handle
    )

    if err != KERN_SUCCESS {
      NSLog(
        "Failed to register notification for touchpad attach: %x, will not handle newly attached devices",
        err)
      IONotificationPortDestroy(port)
      return  // this `return` was not previously here. But it's only logical to have it.
    }

    while case let ioService = IOIteratorNext(handle), ioService != 0 {
      do { IOObjectRelease(ioService) }
    }

    CGDisplayRegisterReconfigurationCallback(
      displayReconfigurationCallback,
      Unmanaged.passUnretained(self).toOpaque()
    )

    registerMouseCallback()
  }

  /// Schedule listeners to be restarted. If a restart is pending, delay it.
  func scheduleRestart(_ delay: TimeInterval) {
    restartTimer?.invalidate()
    restartTimer = Timer.scheduledTimer(
      withTimeInterval: Controller.immediateRestart ? 0 : delay, repeats: false
    ) { [weak self] _ in
      guard let self = self else { return }
      DispatchQueue.main.async {
        self.restartListeners()
      }
    }
  }

  /// Callback for system wake up.
  /// Can be tested by entering `pmset sleepnow` in the Terminal
  @objc func receiveWakeNote(_ note: Notification) {
    NSLog("System woke up, restarting in \(Controller.wakeRestartTimeout)...")
    scheduleRestart(Controller.wakeRestartTimeout)
  }

  func getClickMode() -> Bool {
    return needToClick
  }

  func setMode(_ click: Bool) {
    UserDefaults.standard.set(click, forKey: MiddleClickConfig.needClickKey)
    needToClick = click
  }

  func resetClickMode() {
    UserDefaults.standard.removeObject(forKey: MiddleClickConfig.needClickKey)
    needToClick = getIsSystemTapToClickDisabled()
  }

  func restartListeners() {
    NSLog("Restarting app functionality...")
    stopUnstableListeners()
    startUnstableListeners()
  }

  func startUnstableListeners() {
    NSLog("Starting unstable listeners...")
    registerTouchCallback()
    registerMouseCallback()
  }

  func stopUnstableListeners() {
    NSLog("Stopping unstable listeners...")
    unregisterTouchCallback()
    unregisterMouseCallback()
  }

  func registerMouseCallback() {
    let eventMask = CGEventMask.from(.leftMouseDown, .leftMouseUp, .rightMouseDown, .rightMouseUp)
    currentEventTap = CGEvent.tapCreate(
      tap: .cghidEventTap, place: .headInsertEventTap, options: .defaultTap,
      eventsOfInterest: eventMask, callback: mouseCallback, userInfo: nil)

    if let tap = currentEventTap {
      currentRunLoopSource = CFMachPortCreateRunLoopSource(
        kCFAllocatorDefault, tap, 0)
      CFRunLoopAddSource(
        CFRunLoopGetCurrent(), currentRunLoopSource, .commonModes)
      CGEvent.tapEnable(tap: tap, enable: true)
    } else {
      NSLog("Couldn't create event tap! Check accessibility permissions.")
      UserDefaults.standard.set(true, forKey: "NSStatusItem Visible Item-0")
      scheduleRestart(5)
    }
  }

  private func unregisterMouseCallback() {
    // Disable the event tap first
    if let eventTap = currentEventTap {
      CGEvent.tapEnable(tap: eventTap, enable: false)
    }

    // Remove and release the run loop source
    if let runLoopSource = currentRunLoopSource {
      CFRunLoopRemoveSource(CFRunLoopGetCurrent(), runLoopSource, .commonModes)
      currentRunLoopSource = nil
    }

    // Release the event tap
    currentEventTap = nil
  }

  func registerTouchCallback() {
    currentDeviceList =
      (MTDeviceCreateList()?.takeUnretainedValue() as? [MTDevice]) ?? []

    currentDeviceList.forEach { registerMTDeviceCallback($0, touchCallback) }
  }

  func unregisterTouchCallback() {
    currentDeviceList.forEach { unregisterMTDeviceCallback($0, touchCallback) }
    currentDeviceList.removeAll()
  }

  func getIsSystemTapToClickDisabled() -> Bool {
    let output = runCommand(
      "defaults read com.apple.driver.AppleBluetoothMultitouch.trackpad Clicking"
    )
    return output == "0\n"
  }

  func runCommand(_ command: String) -> String {
    let pipe = Pipe()
    let task = Process()
    task.launchPath = "/bin/sh"
    task.arguments = ["-c", command]
    task.standardOutput = pipe

    task.launch()

    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    return String(data: data, encoding: .utf8) ?? ""
  }

}

@MainActor let mouseCallback: CGEventTapCallBack = {
  proxy, type, event, refcon in
  if needToClick {
    if threeDown && (type == .leftMouseDown || type == .rightMouseDown) {
      wasThreeDown = true
      event.type = .otherMouseDown
      event.setIntegerValueField(.mouseEventButtonNumber, value: kCGMouseButtonCenter)
      threeDown = false
    }

    if wasThreeDown && (type == .leftMouseUp || type == .rightMouseUp) {
      wasThreeDown = false
      event.type = .otherMouseUp
      event.setIntegerValueField(.mouseEventButtonNumber, value: kCGMouseButtonCenter)
    }
  }

  return Unmanaged.passUnretained(event)
}

@MainActor func touchCallback(
  device: Int32, data: UnsafeMutablePointer<Finger>?, nFingers: Int32,
  timestamp: Double, frame: Int32
) -> Int32 {
  if needToClick {
    threeDown =
      allowMoreFingers ? nFingers >= fingersQua : nFingers == fingersQua
    return 0
  }

  if nFingers == 0 {
    if let startTime = touchStartTime {
      let elapsedTime = -startTime.timeIntervalSinceNow
      touchStartTime = nil

      if middleClickX + middleClickY > 0 && elapsedTime <= maxTimeDelta {
        let delta =
          abs(middleClickX - middleClickX2)
          + abs(middleClickY - middleClickY2)
        if delta < maxDistanceDelta {
          emulateMiddleClick()
        }
      }
    }
  } else if nFingers > 0 && touchStartTime == nil {
    touchStartTime = Date()
    maybeMiddleClick = true
    middleClickX = 0.0
    middleClickY = 0.0
  } else if maybeMiddleClick {
    let elapsedTime = -(touchStartTime?.timeIntervalSinceNow ?? 0)
    if elapsedTime > maxTimeDelta {
      maybeMiddleClick = false
    }
  }

  if !allowMoreFingers && nFingers > fingersQua {
    maybeMiddleClick = false
    middleClickX = 0.0
    middleClickY = 0.0
  }

  if allowMoreFingers ? nFingers >= fingersQua : nFingers == fingersQua {
    if maybeMiddleClick {
      //      middleClickX = 0.0 // TODO check if setting to 0 is necessary here
      //      middleClickY = 0.0
      for i in 0..<fingersQua {
        if let fingerData = data?.advanced(by: Int(i)).pointee {
          let pos = fingerData.normalized.pos
          middleClickX += pos.x
          middleClickY += pos.y
        }
      }
      middleClickX2 = middleClickX
      middleClickY2 = middleClickY
      maybeMiddleClick = false
    } else {
      middleClickX2 = 0.0
      middleClickY2 = 0.0
      for i in 0..<fingersQua {
        if let fingerData = data?.advanced(by: Int(i)).pointee {
          let pos = fingerData.normalized.pos
          middleClickX2 += pos.x
          middleClickY2 += pos.y
        }
      }
    }
  }

  return 0
}

/// TODO? is this restart necessary? I don't see any changes when it's removed, but keep in mind I've only spent 5 minutes testing different app and system states
@MainActor public let displayReconfigurationCallback:
  CGDisplayReconfigurationCallBack = { display, flags, userData in
    if flags.contains(.setModeFlag) || flags.contains(.addFlag)
      || flags.contains(.removeFlag) || flags.contains(.disabledFlag)
    {
      print("Display reconfigured, restarting...")
      let controller = Unmanaged<Controller>.fromOpaque(userData!)
        .takeUnretainedValue()
      controller.scheduleRestart(2)
    }
  }

public func emulateMiddleClick() {
  // get the current pointer location
  let location = CGEvent(source: nil)?.location ?? .zero
  let buttonType: CGMouseButton = .center

  postMouseEvent(type: .otherMouseDown, button: buttonType, location: location)
  postMouseEvent(type: .otherMouseUp, button: buttonType, location: location)
}

public func postMouseEvent(
  type: CGEventType, button: CGMouseButton, location: CGPoint
) {
  if let event = CGEvent(
    mouseEventSource: nil, mouseType: type, mouseCursorPosition: location,
    mouseButton: button)
  {
    event.post(tap: .cghidEventTap)
  }
}

@MainActor let multitouchDeviceAddedCallback: IOServiceMatchingCallback = {
  (userData, iterator) in
  while case let ioService = IOIteratorNext(iterator), ioService != 0 {
    do { IOObjectRelease(ioService) }
  }

  let controller = Unmanaged<Controller>.fromOpaque(userData!)
    .takeUnretainedValue()
  NSLog("Multitouch device added, restarting...")
  controller.scheduleRestart(2)
}

func registerMTDeviceCallback(
  _ device: MTDevice, _ callback: @escaping MTContactCallbackFunction
) {
  MTRegisterContactFrameCallback(device, callback)
  MTDeviceStart(device, 0)
}

func unregisterMTDeviceCallback(
  _ device: MTDevice, _ callback: @escaping MTContactCallbackFunction
) {
  MTUnregisterContactFrameCallback(device, callback)
  MTDeviceStop(device)
  MTDeviceRelease(device)
}

extension CGEventMask {
  static func from(_ types: CGEventType...) -> Self {
    var mask = 0

    for type in types {
      mask |= (1 << type.rawValue)
    }

    return Self(mask)
  }
}
